include/seqan3/core/debug_stream/range.hpp

```diff
@@ -68,80 +69,162 @@ constexpr bool reference_type_is_streamable_v<rng_t, char_t> = true;
+ * This printer excludes recursive ranges, such as entities from the std::filesystem library.
+ *
+ * \tparam rng_t Type of the range to be printed; must model std::ranges::input_range and be non-recursive.
+ * \ingroup core_debug_stream
  */
-template <typename char_t, std::ranges::input_range rng_t>
-inline debug_stream_type<char_t> & operator<<(debug_stream_type<char_t> & s, rng_t && r)
-    requires detail::debug_stream_range_guard<rng_t>
+template <typename rng_t>
+    requires std::ranges::input_range<rng_t> && nonrecursive_range<rng_t>
+struct input_range_printer<rng_t>
 {
-    static_assert(detail::reference_type_is_streamable_v<rng_t, char_t>,
-                  "The reference type of the passed range cannot be streamed into the debug_stream.");
-
-    s << '[';
-    auto b = std::ranges::begin(r);
-    auto e = std::ranges::end(r);
-    if (b != e)
+    /*!
+     * \brief Prints the elements of a sequence to an output stream.
+     *
+     * \tparam stream_t The type of the output stream.
+     * \tparam range_t The type of the range to be printed; must be the same type as `rng_t` after .
+     *
+     * \param[in,out] stream The output stream to print to.
+     * \param[in] r The range to be printed.
+     */
+    template <typename stream_t, typename range_t>
+        requires std::same_as<std::remove_cvref_t<range_t>, std::remove_cvref_t<rng_t>>
+    constexpr void operator()(stream_t & stream, range_t && r) const
     {
-        s << *b;
-        ++b;
+        stream << '[';
+        auto first = std::ranges::begin(r);
+        auto last = std::ranges::end(r);
+        if (first != last)
+        {
+            stream << *first;
+            ++first;
+        }
+        while (first != last)
+        {
+            stream << ',';
+            stream << *first;
+            ++first;
+        }
+        stream << ']';
     }
-    while (b != e)
+};
+
+/*!
+ * \brief A printer for (biological) sequences.
+ *
+ * The (biological) sequence (except for ranges over unsigned integers) is printed just as if it were a string, i.e.
+ * <tt>std::vector<dna4>{'C'_dna4, 'G'_dna4, 'A'_dna4}</tt> is printed as "CGA".
+ *
+ * \tparam sequence_t The type of the sequence to be printed; must model seqan3::sequence
+ * \ingroup core_debug_stream
+ */
+template <typename sequence_t>
+    requires sequence<sequence_t>
+struct sequence_printer<sequence_t>
+{
+    /*!
+     * \brief Prints the elements of a sequence to an output stream.
+     *
+     * \tparam stream_t The type of the output stream.
+     * \tparam range_t The type of the sequence to be printed; must be the same type as `sequence_t`.
+     *
+     * \param[in,out] stream The output stream to print to.
+     * \param[in] sequence The sequence to be printed.
+     */
+    template <typename stream_t, typename range_t>
+        requires std::same_as<std::remove_cvref_t<range_t>, std::remove_cvref_t<sequence_t>>
+    constexpr void operator()(stream_t & stream, range_t && sequence) const
```

Comment:
What was the issue here that we need a universal reference?

(The "fix perfect forwarding issue in range printer" commit)